### DISCUSSION TITLE
Allow indexing arrays by a range

### DISCUSSION THREAD

**Discussion by SleepingShell** [noir-lang/noir#3139] [Ideas]
### Problem

Enable indexing of arrays by ranges such as in [Rust](https://doc.rust-lang.org/std/ops/struct.Range.html).

It would also be useful to assign with ranges.

### Happy Case

```rust
let x = [1,2,3, 4];
assert(x[0..2] == [1,2]);
assert(x[1..3] == [2,3]);
assert(x[2..] == [3,4]);
```

Assignment:
```rust
let mut x = [1,2,3];
x[0..2] = [4,4,4];
assert(x == [4,4,4]);
```

### Would you like to submit a PR for this Issue?

No

**Comment by Savio-Sou** [comment]
@jfecher any thoughts?

**Comment by jfecher** [comment]
It's possible to implement without introducing too many new features but it'd likely require a copy of each index rather than a view into them since arrays are immutable internally. So you could easily create slices this way but mutating those slices afterwards wouldn't affect the original array.
